### PR TITLE
create build for tokens

### PR DESCRIPTION
**Pull request by wibjorn**
task-378375

This PR creates a new package in the repo for the express purpose of taking our token files and transfering them into json and js format. This PR also includes a small proof of concept of how tokens can be integrated into the `/site`'s build system.

## Testing

1. Checkout branch.
2. Yarn install.
3. See below.

### Command to test

- `yarn build tokens` - runs the code to read our scss and copy it into json and js. You should see the output in the `/plugins/tokens/dist` folder.
- `yarn start` - should work and you should be able to view the typography page with tokens interpolated (though it isn't pretty yet!)
- after the above visit: http://localhost:1111/tokens/typography.html
- `yarn build` - should also work.

## Additional information

This package relies on `sass-extract` for it's parsing of Scss. Unfortunately `Sass-extract` relies on `node-sass` (Atlas will continue to use dart-sass). In order to cirumvent problems, it was necessary to include the tokens code into a package separate from /css.

The implications of this are as follows:
1.) The syntax for the `/css/tokens` folder must use `@import` as node-sass does not support @forward or @use.
2.) We cannot easily use this package to further tokenize files outside the /tokens folder.

Additionally, the code `/plugins/tokens` makes some changes to the structure returned by `extract-sass`. The reason for this is to provide great interop. Mustache for example, doesn't accessing/displaying keys in objects, so I've included both a `.list ([])` and `.map ({})` property on tokens. This means we'll get the quick access of variables with `.map` as well easy iteration with `.list`. It also means the files are larger than they would otherwise be.

### CODE REVIEW

File: plugins/tokens/index.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		});
+}
+
+/**
+ *
+ * @param {import('./types').SassConvertRendered} rendered
+ * @param { {[key:string]: any}} tokens
+ */
+function writeVariables(rendered, tokens) {
+	for (const name in rendered.vars.global) {
```
Comment (jdanyow) [inline plugins/tokens/index.js:51]: super nit, feel free to resolve, I think you can use 
```js
for (const [name, rule] of Object.entries(rendered.vars.global)) {
```

File: plugins/parcel-transformer-markdown-html/lib/index.js [hunk 2]
```diff
@@ -68,6 +68,21 @@ module.exports = new Transformer({
 					: Promise.resolve(null)
 			]);
 
+			let tokens;
+			// we've specified a tokens file to load from @atlas-tokens
+			if (attributes.token) {
+				try {
+					const jsm = require(`@microsoft/atlas-tokens/dist/${attributes.token}.js`);
```
Comment (jdanyow) [inline plugins/parcel-transformer-markdown-html/lib/index.js:75]: I think when the main is updated you'll be able to use `const tokens = require('@microsoft/atlas-tokens')`

File: plugins/tokens/index.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			return tokens;
+		}, {});
+	return parents;
+}
+
+function writeJsToken(json) {
+	return `'use strict';
+Object.defineProperty(exports, "__esModule", { value: true });
+exports.default = ${json};`;
+}
```
Comment (jdanyow) [inline plugins/tokens/index.js:96]: I don't think this is needed

File: plugins/tokens/index.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
+//@ts-check
+const sassExtract = require('sass-extract');
+const glob = require('fast-glob');
+const fs = require('fs-extra');
+const path = require('path');
+const { getFinalFilePart, convertSassVariable } = require('./conversation');
+
+const distFolder = path.join(process.cwd(), '/dist');
```
Comment (jdanyow) [inline plugins/tokens/index.js:8]: I don't think this is needed

File: plugins/tokens/index.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		.then(async rendered => {
+			const tokens = collectTokenSources(rendered);
+			writeVariables(rendered, tokens);
+
+			Object.keys(tokens).map(token => {
+				const outfileStem = path.join(distFolder, token);
+				const json = JSON.stringify(tokens[token]);
+				// writing json output
+				fs.writeFile(outfileStem + '.json', json);
+				fs.writeFile(outfileStem + '.js', writeJsToken(json));
```
Comment (jdanyow) [inline plugins/tokens/index.js:37]: I don't think this is needed

File: plugins/tokens/index.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if (entries.length !== 1) {
+		throw new Error(
+			`Error with entries. There should only be a single entry. Got ${entries.length}.`
+		);
+	}
+
+	await fs.mkdir(distFolder);
+
+	sassExtract
+		.render({
```
Comment (jdanyow) [inline plugins/tokens/index.js:25]: await the render?

File: plugins/tokens/index.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	// Global variables play better when we just import a single index file
+	const entries = await glob(['../../css/src/tokens/index.scss'], {});
+
+	if (entries.length !== 1) {
+		throw new Error(
+			`Error with entries. There should only be a single entry. Got ${entries.length}.`
+		);
+	}
+
+	await fs.mkdir(distFolder);
```
Comment (jdanyow) [inline plugins/tokens/index.js:22]: await fs.ensureDir('./dist');

File: plugins/tokens/index.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		})
+		.then(async rendered => {
+			const tokens = collectTokenSources(rendered);
+			writeVariables(rendered, tokens);
+
+			Object.keys(tokens).map(token => {
+				const outfileStem = path.join(distFolder, token);
+				const json = JSON.stringify(tokens[token]);
+				// writing json output
+				fs.writeFile(outfileStem + '.json', json);
```
Comment (jdanyow) [inline plugins/tokens/index.js:36]: super nit- fs-extra has writeJson `await fs.writeJson('./dist/index.json', tokens)`

File: plugins/tokens/package.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	"homepage": "https://github.com/microsoft/atlas-design",
+	"keywords": [
+		"stylelint",
+		"stylelint-config",
+		"atlas",
+		"atlas-css"
+	],
+	"author": "Microsoft Corporation",
+	"license": "MIT",
+	"main": "index.js",
```
[the remark this answers is not in the document]
Comment (jdanyow) [reply plugins/tokens/package.json:15]: (with the assumption there's just one json output, might not be valid)

**Review by jdanyow — CHANGES_REQUESTED** [review]
looks good, left some comments centered around the dist/main output

File: plugins/tokens/conversation.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
+//@ts-check
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/conversation.js:1]: 🤦‍♂️, 🤔,🤨,🥱, 👍

File: plugins/tokens/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/package.json:15]: We could output a dist/index.js that could point to the various <token>.json files, but right now there isn't one.

File: plugins/tokens/color.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
+//@ts-check
+const colorConverter = require('color-convert');
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/color.js]: removed

File: plugins/tokens/conversation.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/conversation.js]: done

File: plugins/tokens/index.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
+//@ts-check
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/index.js]: done

File: plugins/tokens/index.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+const path = require('path');
+const { getFinalFilePart, convertSassVariable } = require('./conversation');
+
+const distFolder = path.join(process.cwd(), '/dist');
+
+createTokens(distFolder);
+
+async function createTokens(distFolder) {
+	// Global variables play better when we just import a single index file
+	const entries = await glob(['../../css/src/tokens/index.scss'], {});
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/index.js]: good call! done

File: plugins/tokens/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/package.json]: after condensing the input significantly, we're creating only one json file.

File: plugins/tokens/types.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+ * @typedef {'SassList' | 'SassString' | 'SassNumber' | 'SassColor' | 'SassObject'} SassType
+ */
+
+/**
+ * @typedef {SassStringRule | SassNumberRule | SassColorRule | SassListRule | SassMapRule} SassRule
+ */
+
+/**
+ * @typedef RuleDeclaration
+ ?* @property {string} expression;
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/types.js:13]: it's typescript now

File: plugins/tokens/types.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
+export function thisAllowsJSDocTypesToBeImported() {}
```
[the remark this answers is not in the document]
Comment (wibjorn) [reply plugins/tokens/types.js:1]: I had no idea it was so easy! Done

File: plugins/tokens/index.js (hunk 4 above)
Comment (wibjorn) [reply plugins/tokens/index.js]: removed

File: plugins/tokens/index.js (hunk 6 above)
Comment (wibjorn) [reply plugins/tokens/index.js]: done

File: plugins/tokens/index.js (hunk 7 above)
Comment (wibjorn) [reply plugins/tokens/index.js]: done

File: plugins/tokens/index.js (hunk 8 above)
Comment (wibjorn) [reply plugins/tokens/index.js]: done, though with quicktype needing a string it doesn't matter much

File: plugins/tokens/index.js (hunk 5 above)
Comment (wibjorn) [reply plugins/tokens/index.js]: removed

File: plugins/tokens/index.js (hunk 3 above)
Comment (wibjorn) [reply plugins/tokens/index.js:123]: no more js output

File: package.json [hunk 16]
```diff
@@ -16,7 +16,11 @@
 		"toc": "yarn workspace @microsoft/atlas-site toc",
 		"prettier": "prettier --write \"**/*.{scss,ts,js,json,yml,md}\"",
 		"prettier-check": "prettier --check \"**/*.{scss,ts,js,json,yml}\" --loglevel debug",
-		"pretty-quick": "pretty-quick --staged --pattern \"**/*.{scss,ts,js,json,yml}\""
+		"pretty-quick": "pretty-quick --staged --pattern \"**/*.{scss,ts,js,json,yml}\"",
+		"prestart": "yarn run build:tokens",
+		"build:tokens": "yarn workspace @microsoft/atlas-css tokens",
+		"postinstall": "yarn run build:tokens",
```
Comment (jdanyow) [inline package.json:22]: needed?

**Review by jdanyow — APPROVED** [review]

**Review by jdanyow — APPROVED** [review]

**Pull request merged**